### PR TITLE
MGMT-6806 Support pushing remote debug image to local registry on loc…

### PR DESCRIPTION
**Pull request by michaellevy101**
…alhost

### CODE REVIEW

File: discovery-infra/test_infra/utils/utils.py [hunk 1]
```diff
@@ -492,9 +492,10 @@ def get_local_assisted_service_url(namespace, service, deploy_target):
     else:
         # default deploy target is minikube
         log.info('Getting minikube %s URL in %s namespace', service, namespace)
-        url, _, _ = run_command(
-            f'minikube -n {namespace} service {service} --url'
-        )
+        ip, _, _ = run_command("minikube ip")
+        # resolve the service node port form the service internal port (Support multiple ports per service).
+        port, _, _ = run_command(f'kubectl get svc assisted-service -n {namespace} -o=jsonpath="{{.spec.ports[?(@.port==8090)].nodePort}}"')
```
Comment (eliorerz) [inline discovery-infra/test_infra/utils/utils.py:497]: Why it's part of the code? Why not passing it as variables?

Comment (michaellevy101) [reply discovery-infra/test_infra/utils/utils.py:497]: because minikube is a special case in this function

Comment (eliorerz) [reply discovery-infra/test_infra/utils/utils.py:497]: @YuviGold what do you think?

Comment (YuviGold) [reply discovery-infra/test_infra/utils/utils.py:497]: Could we just get the service external IP? when we work with the tunnel, the cluster has a load balancer and the service should be exposed. This is how we work with assisted-service when waiting for it to be healthy
https://github.com/openshift/assisted-service/blob/4f07ab7ea86cdf723d04b9e8a70e30b8e0f627b4/tools/utils.py#L116-L140
This way we won't need to have the minikube dependency.

File: scripts/deploy_assisted_service.sh [hunk 2]
```diff
@@ -131,6 +137,8 @@ else
     wait_for_url_and_run ${SERVICE_BASE_URL} "spawn_port_forwarding_command $SERVICE_NAME $SERVICE_PORT $NAMESPACE $NAMESPACE_INDEX $KUBECONFIG minikube undeclared $SERVICE_INTERNAL_PORT"
 
     if [ "${DEBUG_SERVICE}" == "true" ]; then
+        # Close the minikube registry service port forwarding
+        pkill -f "port-forward"
```
Comment (YuviGold) [inline scripts/deploy_assisted_service.sh:141]: Why would we need to do this?

File: scripts/deploy_assisted_service.sh [hunk 3]
```diff
@@ -117,9 +117,15 @@ else
     print_log "Updating assisted_service params"
 
     if [ "${DEBUG_SERVICE}" == "true" ]; then
-        print_log " Patching assisted service image with a debuggable code "
-        (cd assisted-service/ && skipper --env-file ../skipper.env make update-local-image -e LOCAL_ASSISTED_ORG=$(minikube ip):5000 -e CONTAINER_BUILD_EXTRA_PARAMS="--cgroup-manager=cgroupfs --storage-driver=vfs --events-backend=file")
+        # Forward the minikube registry addon k8s service to the host to push the debug image using localhost:5000
+        kubectl port-forward service/registry -n kube-system 5000:80 > /dev/null 2>&1 &
```
Comment (YuviGold) [inline scripts/deploy_assisted_service.sh:121]: Work with xinetd like we do for the rest of the port-forwarding with `spawn_port_forwarding_command`.

File: scripts/deploy_assisted_service.sh [hunk 4]
```diff
@@ -117,9 +117,15 @@ else
...
+        # Set the local registry to the minikube registry (used by the assisted-service update-local-image target)
+        export LOCAL_ASSISTED_ORG=localhost:5000
+        print_log "Patching assisted service image with a debuggable code "
+        (cd assisted-service/ && skipper --env-file ../skipper.env make update-local-image -e CONTAINER_BUILD_EXTRA_PARAMS="--cgroup-manager=cgroupfs --storage-driver=vfs --events-backend=file")
```
Comment (YuviGold) [inline scripts/deploy_assisted_service.sh:125]: export `CONTAINER_BUILD_EXTRA_PARAMS` like you did for `LOCAL_ASSISTED_ORG`.

File: scripts/deploy_assisted_service.sh [hunk 5]
```diff
@@ -117,9 +117,15 @@ else
...
         DEBUG_DEPLOY_AI_PARAMS="REPLICAS_COUNT=1"
+        # Override the SERVICE environment variable with the local registry debug image
+        export SERVICE="$LOCAL_ASSISTED_ORG/assisted-service:latest"
```
Comment (YuviGold) [inline scripts/deploy_assisted_service.sh:128]: Use curly parenthesis when working with bash variables.

File: scripts/deploy_assisted_service.sh (hunk 2 above)
Comment (michaellevy101) [reply scripts/deploy_assisted_service.sh:141]: cleanup

File: scripts/deploy_assisted_service.sh (hunk 3 above)
Comment (michaellevy101) [reply scripts/deploy_assisted_service.sh:121]: why? It shouldn't stay open

File: scripts/deploy_assisted_service.sh (hunk 4 above)
Comment (michaellevy101) [reply scripts/deploy_assisted_service.sh:125]: It used only once

File: discovery-infra/test_infra/utils/utils.py (hunk 1 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/utils/utils.py:497]: We should push to localhost, pushing to localhost doesn't require ssl...

Comment (YuviGold) [reply discovery-infra/test_infra/utils/utils.py:497]: Sorry, I don't understand your answer...

File: scripts/deploy_assisted_service.sh (hunk 3 above)
Comment (YuviGold) [reply scripts/deploy_assisted_service.sh:121]: To be aligned with all the rest of the post-forwarding solutions.

File: scripts/deploy_assisted_service.sh (hunk 4 above)
Comment (YuviGold) [reply scripts/deploy_assisted_service.sh:125]: I just think it would look nicer.. but not critical.

File: scripts/deploy_assisted_service.sh (hunk 2 above)
Comment (YuviGold) [reply scripts/deploy_assisted_service.sh:141]: Wouldn't that kill other port-fowarding jobs running on the machine?

File: scripts/deploy_assisted_service.sh (hunk 3 above)
Comment (michaellevy101) [reply scripts/deploy_assisted_service.sh]: anyway added to the xinetd

File: scripts/deploy_assisted_service.sh (hunk 2 above)
Comment (michaellevy101) [reply scripts/deploy_assisted_service.sh]: After adding the registry to the xinetd,  it's not relevant any more

File: discovery-infra/test_infra/utils/utils.py (hunk 1 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/utils/utils.py:497]: This code trying to resolve the assisted-service ip from the  Load balancer service....
minikube tunnel is like a node port service (minikube is a single node cluster) but the port is different.
Here i'm resolving the external port from the service internal port which is 8090.
please see the table below:

NAMESPACE            NAME               TYPE           CLUSTER-IP       EXTERNAL-IP      PORT(S)                          AGE
assisted-installer   assisted-service   LoadBalancer   10.104.184.193   10.104.184.193   8090:31951/TCP,40000:32463/TCP   44h
assisted-installer   object-store       ClusterIP      10.103.27.13     <none>           9000/TCP                         44h
assisted-installer   ocp-metal-ui       LoadBalancer   10.99.62.228     10.99.62.228     80:30370/TCP                     44h
assisted-installer   postgres           LoadBalancer   10.106.135.23    10.106.135.23    5432:32057/TCP                   44h
default              kubernetes         ClusterIP      10.96.0.1        <none>           443/TCP                          44h
kube-system          kube-dns           ClusterIP      10.96.0.10       <none>           53/UDP,53/TCP,9153/TCP           44h
kube-system          registry           ClusterIP      10.107.138.33    <none>           80/TCP,443/TCP                   44h

Comment (YuviGold) [reply discovery-infra/test_infra/utils/utils.py:497]: I agree. I just want you to replace the `minikube ip` :)  So the solution would work with other clusters as well.

Comment (michaellevy101) [reply discovery-infra/test_infra/utils/utils.py:497]: how can we get the external ip...

Comment (YuviGold) [reply discovery-infra/test_infra/utils/utils.py:497]: I linked above how we do it already in assisted-service deployment scripts, when waiting for the service to be healthy. :)

Comment (michaellevy101) [reply discovery-infra/test_infra/utils/utils.py:497]: do you suggest to extract it from: "kubectl get nodes -o=jsonpath={.items[0].status.addresses[0].address}" this is not a general solution...

**Review by YuviGold — APPROVED** [review]
